agent/containers_dockercli.go

```diff
@@ -0,0 +1,188 @@
+package agent
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coder/coder/v2/codersdk"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/xerrors"
+)
+
+// dockerCLIContainerLister is a ContainerLister that lists containers using the docker CLI
+type dockerCLIContainerLister struct{}
+
+var _ ContainerLister = &dockerCLIContainerLister{}
+
+func (*dockerCLIContainerLister) List(ctx context.Context) ([]codersdk.WorkspaceAgentContainer, error) {
+	var buf bytes.Buffer
+	// List all container IDs, one per line, with no truncation
+	cmd := exec.CommandContext(ctx, "docker", "ps", "--all", "--quiet", "--no-trunc")
+	cmd.Stdout = &buf
+	if err := cmd.Run(); err != nil {
+		return nil, xerrors.Errorf("run docker ps: %w", err)
+	}
+
+	ids := make([]string, 0)
+	for _, line := range strings.Split(buf.String(), "\n") {
```

Comment:
Nit: we can preallocate here, assuming all lines are relevant
Most lines in `docker ps` output are relevant.

```suggestion
        lines := strings.Split(buf.String(), "\n")
	ids := make([]string, 0, len(lines))
	for _, line := range lines {
```